src/main/java/net/replaceitem/integratedcircuit/client/gui/IntegratedCircuitScreen.java

```diff
@@ -212,120 +282,145 @@ public static void renderComponentTexture(DrawContext drawContext, Identifier co
         renderPartialTexture(drawContext, texture, componentX, componentY, x, y, textureW, textureH, rot, color, 0, 0, textureW, textureH);
     }
 
     private static void renderPartialTexture(DrawContext drawContext, Identifier texture, int componentX, int componentY, int x, int y, int textureW, int textureH, int rot, int color, int u, int v, int w, int h) {
         drawContext.getMatrices().push();
-        drawContext.getMatrices().translate(componentX+8, componentY+8, 0);
-        drawContext.getMatrices().multiply(RotationAxis.POSITIVE_Z.rotation((float) (rot*Math.PI*0.5)));
+        drawContext.getMatrices().translate(componentX + 8, componentY + 8, 0);
+        drawContext.getMatrices().multiply(RotationAxis.POSITIVE_Z.rotation((float) (rot * Math.PI * 0.5)));
         drawContext.getMatrices().translate(-8, -8, 0);
         RenderSystem.enableBlend();
         drawContext.drawTexture(RenderLayer::getGuiTextured, texture, x, y, u, v, w, h, textureW, textureH, color);
         drawContext.getMatrices().pop();
     }
 
-    private void selectPalette(int index) {
-        if(index < 0 || index >= componentButtons.size()) return;
-        if(selectedComponentSlot >= 0 && selectedComponentSlot < componentButtons.size()) {
-            componentButtons.get(selectedComponentSlot).setSelected(false);
-        }
-        selectedComponentSlot = index;
-        componentButtons.get(index).setSelected(true);
-        Component component = componentButtons.get(index).getComponent();
-        this.cursorState = component.getDefaultState();
-        if(component instanceof FacingComponent) this.cursorState = this.cursorState.with(FacingComponent.FACING, this.cursorRotation);
-    }
-
-    private void deselectPalette() {
-        if(selectedComponentSlot >= 0 && selectedComponentSlot < componentButtons.size()) {
-            componentButtons.get(selectedComponentSlot).setSelected(false);
-        }
-        this.cursorState = null;
-        selectedComponentSlot = -1;
-    }
-    
-    private void pickPalette(Component component) {
-        for (int i = 0; i < componentButtons.size(); i++) {
-            if(componentButtons.get(i).getComponent() == component) {
-                selectPalette(i);
-            }
-        }
+    @Override
+    public <T extends Element & Drawable & Selectable> T addDrawableChild(T drawableElement) {
+        return super.addDrawableChild(drawableElement);
     }
 
     @Override
     public boolean mouseClicked(double mouseX, double mouseY, int button) {
-        if(this.client == null) return false;
+        if (this.client == null)
+            return false;
+
         ComponentPos clickedPos = getComponentPosAt((int) mouseX, (int) mouseY);
 
+        if (customNameTextField.isFocused() && !customNameTextField.isMouseOver(mouseX, mouseY)) {
```

Comment:
You can override the setFocused method of TextFieldWidget when constructing it and use that for detecting when it loses focus. Then you wouldn't need the extra checks in keyPressed or mouseClicked